crates/adapters/src/transport/s3/mod.rs

```
@@ -0,0 +1,482 @@
+        tokio::spawn(async move {
+            let mut continuation_token = None;
+            loop {
+                let (objects_to_fetch, next_token): (Vec<String>, Option<String>) = match &config
+                    .read_strategy
+                {
+                    ReadStrategy::Prefix { prefix } => {
+                        let result = client
+                            .get_object_keys(&config.bucket_name, prefix, continuation_token.take())
+                            .await;
+                        if result.is_err() {
+                            error!("Could not fetch object keys {result:?}. Bailing.");
+                            break;
+                        }
+                        result.unwrap()
+                    }
+                    ReadStrategy::SingleKey { key } => (vec![key.clone()], None),
+                };
+                continuation_token = next_token;
+                for key in &objects_to_fetch {
+                    let object = client.get_object(&config.bucket_name, key).await;
+                    tx.send(object).await.expect("Enqueue failed");
+                }
+                if continuation_token.is_none() {
+                    break;
+                }
+            }
+            drop(tx); // We're done. Close the channel.
+        });
+        loop {
+            let state = *receiver.borrow();
+            match state {
+                PipelineState::Paused => {
+                    receiver.changed().await?;
+                }
+                PipelineState::Running => {
+                    tokio::select! {
+                        // If pipeline status has changed, exit the select
+                        _ = receiver.changed() => (),
+                        // Poll the object stream. If `None`, we have processed all objects.
+                        get_obj = rx.recv() => {
+                            if let Some(get_obj) = get_obj {
+                                let mut object = get_obj?;
```

I think we should just report the transport error to the controller and abandon this object, but not kill the whole connector.